### PR TITLE
Strategies: Implement curve 3 pool

### PR DESCRIPTION
**Pull request by facuspagnuolo**

### CODE REVIEW

File: packages/strategies/contracts/curve/Curve3PoolStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+        // Join pool
+        uint256 initialPoolTokenBalance = poolToken.balanceOf(address(this));
+        IERC20(token).approve(address(pool), amount);
+        pool.add_liquidity(amounts, minAmountOut);
+        uint256 finalPoolTokenBalance = poolToken.balanceOf(address(this));
+        value = finalPoolTokenBalance - initialPoolTokenBalance;
+
+        // Stake pool tokens
+        poolToken.approve(address(gauge), value);
```
Comment (dmf7z) [inline packages/strategies/contracts/curve/Curve3PoolStrategy.sol:154]:
```suggestion
        poolToken.approve(address(gauge), finalPoolTokenBalance);
```
Just to make sure that all pool tokens are staked

File: packages/strategies/contracts/curve/Curve3PoolStrategy.sol [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        gauge.deposit(value);
```
Comment (dmf7z) [inline packages/strategies/contracts/curve/Curve3PoolStrategy.sol:155]:
```suggestion
        gauge.deposit(finalPoolTokenBalance);
```
Again, just to make sure that all pool tokens are staked

File: packages/strategies/contracts/curve/Curve3PoolStrategy.sol [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    function exit(uint256 ratio, uint256 slippage, bytes memory)
+        external
+        override
+        returns (uint256 amount, uint256 value)
+    {
+        require(ratio <= FixedPoint.ONE, 'CURVE_INVALID_RATIO');
+        require(slippage <= FixedPoint.ONE, 'CURVE_INVALID_SLIPPAGE');
+
+        // Unstake pool tokens
+        uint256 stakedPoolTokenBalance = gauge.balanceOf(address(this));
```
Comment (dmf7z) [inline packages/strategies/contracts/curve/Curve3PoolStrategy.sol:174]: We are not taking into account the current balance of pool tokens, which is ok, must be always zero.

**Review by dmf7z — COMMENTED** [review]
Looks very clean! Small notes

File: packages/strategies/contracts/curve/Curve3PoolStrategy.sol [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        require(_gauge.lp_token() == address(_poolToken), 'CURVE_INVALID_POOL_TOKEN');
+
+        token = _token;
+        poolToken = _poolToken;
+        crv = IERC20(_gauge.crv_token());
+        pool = _pool;
+        gauge = _gauge;
+
+        uint256 index = 0;
+        for (uint256 i = 0; i < COINS; i++) if (_pool.coins(i) == _token) index = i;
```
[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply packages/strategies/contracts/curve/Curve3PoolStrategy.sol:85]: totally!

File: packages/strategies/contracts/curve/Curve3PoolStrategy.sol (hunk 1 above)
Comment (facuspagnuolo) [reply packages/strategies/contracts/curve/Curve3PoolStrategy.sol:154]: The problem is that if for some reason the wallet receives from outside or has LPTs, these will be invested as well. I'm saying that's wrong, just pointing it out, what do you think?

File: packages/strategies/contracts/curve/Curve3PoolStrategy.sol (hunk 3 above)
Comment (facuspagnuolo) [reply packages/strategies/contracts/curve/Curve3PoolStrategy.sol:174]: I'm more inclined on not using tokens that were previously held by the wallet, let's say airdrop or simply deposited, wdyt?

File: packages/strategies/contracts/curve/Curve3PoolStrategy.sol [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        uint256 poolTokenPrice = pool.get_virtual_price();
+        uint256 expectedAmountOut = exitPoolTokenAmount.mulUp(poolTokenPrice) / tokenScale;
+        uint256 minAmountOut = expectedAmountOut.mulUp(FixedPoint.ONE - slippage);
+
+        // Exit pool
+        uint256 initialTokenAmount = IERC20(token).balanceOf(address(this));
+        pool.remove_liquidity_one_coin(exitPoolTokenAmount, int128(int256(tokenIndex)), minAmountOut);
+        uint256 finalTokenAmount = IERC20(token).balanceOf(address(this));
+        amount = finalTokenAmount - initialTokenAmount;
+        value = (amount * tokenScale).mulDown(poolTokenPrice);
```
Comment (facuspagnuolo) [inline packages/strategies/contracts/curve/Curve3PoolStrategy.sol:188]: ⚠️ here it should be `exitPoolTokenAmount.mulDown(pool.get_virtual_price())`

File: packages/strategies/contracts/curve/Curve3PoolStrategy.sol (hunk 2 above)
Comment (facuspagnuolo) [reply packages/strategies/contracts/curve/Curve3PoolStrategy.sol:153]: See above

File: packages/strategies/contracts/curve/Curve3PoolStrategy.sol [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    using FixedPoint for uint256;
+
+    // Namespace under which the Wallet is registered in the Mimic Registry
+    bytes32 public constant override NAMESPACE = keccak256('STRATEGY');
+
+    // Curve 3 pool uses 3 tokens
+    uint256 public constant COINS = 3;
+
+    // Curve 3 pool uses 1 reward token
+    uint256 public constant REWARDS = 1;
```
[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply packages/strategies/contracts/curve/Curve3PoolStrategy.sol:41]: See above

File: packages/strategies/contracts/curve/Curve3PoolStrategy.sol (hunk 1 above)
Comment (dmf7z) [reply packages/strategies/contracts/curve/Curve3PoolStrategy.sol:152]: Agreed, in v1 strategies it made sense to invest it all, here it does not.

**Review by defioctopus — APPROVED** [review]
👍

**Pull request merged**